### PR TITLE
Add CorrelationId,DeviceInstallationReport

### PR DESCRIPTION
**Pull request by jerrytrieu**

### CODE REVIEW

File: libats-messaging-datatype/src/main/scala/com/advancedtelematic/libats/messaging_datatype/Messages.scala [hunk 1]
```diff
@@ -134,6 +148,13 @@ object Messages {
   case class DeviceUpdateReport(namespace: Namespace, device: DeviceId, updateId: UpdateId, timestampVersion: Int,
                                 operationResult: Map[EcuSerial, OperationResult], resultCode: Int)
 
+  case class DeviceInstallationReport(namespace: Namespace, device: DeviceId,
+                                      correlationId: CorrelationId,
+                                      result: InstallationResult,
+                                      ecuResults: Map[EcuSerial, EcuInstallationResult],
+                                      report: Option[Json],
+                                      receivedAt: Instant)
```
Comment (jerrytrieu) [inline libats-messaging-datatype/src/main/scala/com/advancedtelematic/libats/messaging_datatype/Messages.scala]: @koshelev @houcros wdyt

File: libats-messaging-datatype/src/main/scala/com/advancedtelematic/libats/messaging_datatype/DataType.scala [hunk 2]
```diff
@@ -69,6 +69,14 @@ object DataType {
     def isSuccess:Boolean = resultCode == 0 || resultCode == 1
   }
 
+  final case class InstallationTarget(target: TargetFilename,
+                                      hashes: Map[HashMethod, Refined[String, ValidChecksum]],
```
Comment (koshelev) [inline libats-messaging-datatype/src/main/scala/com/advancedtelematic/libats/messaging_datatype/DataType.scala]: do we need hash and size? I mean what is a chance they are reported from a device?

File: libats-messaging-datatype/src/main/scala/com/advancedtelematic/libats/messaging_datatype/DataType.scala [hunk 3]
```diff
@@ -69,6 +69,10 @@ object DataType {
     def isSuccess:Boolean = resultCode == 0 || resultCode == 1
   }
 
+  final case class InstallationResult(success: Boolean, code: String, description: String)
+
+  final case class EcuInstallationResult(result: InstallationResult, target: Seq[TargetFilename])
```
Comment (houcros) [inline libats-messaging-datatype/src/main/scala/com/advancedtelematic/libats/messaging_datatype/DataType.scala]: What would you think about making:
`EcuInstallationResult(success: Boolean, code: String, description: String, target: Seq[TargetFilename])`.
Else when we want to parse the ECU results we'll have to go `DeviceInstallationReport -> ecuResults -> EcuInstallationResult -> code`.
I don't know if it's worth the duplication though.

Comment (koshelev) [reply libats-messaging-datatype/src/main/scala/com/advancedtelematic/libats/messaging_datatype/DataType.scala]: I like having `InstallationResult` class

Comment (houcros) [reply libats-messaging-datatype/src/main/scala/com/advancedtelematic/libats/messaging_datatype/DataType.scala]: It's just that `EcuInstallationResult` feels so much like a subclass of `InstallationResult` :) But this is more concise, let's go with it.

Comment (jerrytrieu) [reply libats-messaging-datatype/src/main/scala/com/advancedtelematic/libats/messaging_datatype/DataType.scala]: Renamed to `EcuInstallationReport`

File: libats-messaging-datatype/src/main/scala/com/advancedtelematic/libats/messaging_datatype/DataType.scala [hunk 4]
```diff
@@ -69,6 +69,10 @@ object DataType {
     def isSuccess:Boolean = resultCode == 0 || resultCode == 1
   }
 
+  final case class InstallationResult(success: Boolean, code: String, description: String)
+
+  final case class EcuInstallationReport(result: InstallationResult, target: Seq[String], report: Option[Json])
```
Comment (koshelev) [inline libats-messaging-datatype/src/main/scala/com/advancedtelematic/libats/messaging_datatype/DataType.scala]: `target: Seq[TargetFilename]`?

File: libats/src/main/scala/com/advancedtelematic/libats/data/Urn.scala [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  }
+
+  case class IllegalUrnException(val msg: String) extends Exception(msg) with NoStackTrace
+
+  final case class CorrelationId(resource: String, uuid: UUID) extends Urn {
+    override val nid = CorrelationId.namespaceId
+    override val nss = s"$resource:${uuid.toString}"
+  }
+
+  object CorrelationId {
```
Comment (koshelev) [inline libats/src/main/scala/com/advancedtelematic/libats/data/Urn.scala]: What about 
```scala
sealed trait CorrelationId
final case class CampaignId(value: UUID) extends CorrelationId
final case class MultiTargetUpdate(value: UUID) extends CorrelationId

object CorrelationId {
  private[this] val CorrelationIdRe = """^urn:here-ota:(mtu|campaign):([0-9a-fA-F]{8}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{12})$"""r

  implicit val DecoderInstance: Decoder[CorrelationId] = Decoder.decodeString.emap {
    case CorrelationIdRe("mtu", uuid) =>
      Right(MultiTargetUpdate(UUID.fromString(uuid)))

    case CorrelationIdRe("campaign", uuid) =>
      Right(CampaignId(UUID.fromString(uuid)))
    case x =>
      Left(s"'$x' is invalid correlationId")
  }

  implicit val EncoderInstance: Encoder[CorrelationId] = Encoder.encodeString.contramap {
    case CampaignId(x) => s"urn:here-ota:campaign:$x"
    case MultiTargetUpdate(x) => s"urn:here-ota:mtu:$x"
  }
}
```
I believe it gives us type safe correlation id and models the fact that urn is just a way it is encoded in json

Comment (jerrytrieu) [reply libats/src/main/scala/com/advancedtelematic/libats/data/Urn.scala]: Cool!

**Pull request merged**